src/decoder/decoder_utils.ml

```diff
@@ -84,30 +84,29 @@ let channels_converter dst =
           converter := Some (c, _src);
           Audio_converter.Channel_layout.convert c data
 
-let video_scale () =
+let video_scale () : Video.Canvas.Image.t -> Video.Canvas.Image.t =
   let dst_width = Lazy.force Frame.video_width in
   let dst_height = Lazy.force Frame.video_height in
   let video_scale = Video_converter.scaler () ~proportional:true in
   fun img ->
-    let src_width = Video.Image.width img in
-    let src_height = Video.Image.height img in
+    let src_width = Video.Canvas.Image.width img in
+    let src_height = Video.Canvas.Image.height img in
     if (src_width, src_height) = (dst_width, dst_height) then img
     else (
       let scl_width, scl_height =
         if dst_height * src_width < dst_width * src_height then
           (src_width * dst_height / src_height, dst_height)
         else (dst_width, src_height * dst_width / src_width)
       in
-      let img2 = Video.Image.create scl_width scl_height in
+      let img = Video.Canvas.Image.render img in
+      let img2 = Image.YUV420.create scl_width scl_height in
       video_scale img img2;
-      if (scl_width, scl_height) = (dst_width, dst_height) then img2
-      else (
-        let img3 = Video.Image.create dst_width dst_height in
-        Video.Image.blank img3;
-        let x = (dst_width - scl_width) / 2 in
-        let y = (dst_height - scl_height) / 2 in
-        Image.YUV420.add img2 ~x ~y img3;
-        img3))
+      let img2 = Video.Canvas.Image.make img2 in
+      let x = (dst_width - scl_width) / 2 in
+      let y = (dst_height - scl_height) / 2 in
+      img2
+      |> Video.Canvas.Image.translate x y
+      |> Video.Canvas.Image.viewport dst_width dst_height)
```

Comment:
Not a blocker but do you think that, eventually, we could keep the original decoded image and do the scaling on the final output render? This would make it possible to optimize and only write once each pixel on the final image..

Reply:
Yes! The canvas type in ocaml-mm was create with this in mind (although, I don't want to do this right now...)